### PR TITLE
src/rgw: cls async APIs

### PR DESCRIPTION
**Pull request by dorindabassey**
enforcing cls async APIs for all RGW code

Fixes: https://tracker.ceph.com/issues/45622

Signed-off-by: dorindabassey <dorindabassey@gmail.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by dorindabassey** [comment]
Hi @yuvalif please review

**Review by yuvalif — CHANGES_REQUESTED** [review]
after setting the flag in the CMakeList.txt file you should:
- find all compilation errors
- for each one, you should change the API invoked from the synchronous one to the a-synchronous one
- no need to "ifndef" anything

File: src/cls/otp/cls_otp_client.h [hunk 1]
```diff
@@ -28,11 +28,12 @@ namespace rados {
         static int get(librados::ObjectReadOperation *op,
                        librados::IoCtx& ioctx, const std::string& oid,
                        const std::list<std::string> *ids, bool get_all, std::list<otp_info_t> *result);
+#endif
```
Comment (yuvalif) [inline src/cls/otp/cls_otp_client.h:31]: see comment in lines 25-26. these functions should not be called from inside the RGW (they are using IoCtx - hence sync calls).
they should be inside the `ifndef`

File: src/cls/version/cls_version_client.h [hunk 2]
```diff
@@ -23,9 +23,9 @@ void cls_version_read(librados::ObjectReadOperation& op, obj_version *objv);
 
 // these overloads which call io_ctx.operate() or io_ctx.exec() should not be called in the rgw.
 // rgw_rados_operate() should be called after the overloads w/o calls to io_ctx.operate()/exec()
-#ifndef CLS_CLIENT_HIDE_IOCTX
+
```
Comment (yuvalif) [inline src/cls/version/cls_version_client.h:26]: please keep the `ifndef`

File: src/rgw/rgw_gc.cc [hunk 3]
```diff
@@ -203,33 +203,8 @@ int RGWGC::list(int *index, string& marker, uint32_t max, bool expired_only, std
...
-              marker.clear();
-            }
-          }
-        }
-      }
-      if ((objv.ver == 1) && (entries.size() < max - result.size())) {
-        check_queue = true;
-        marker.clear();
-      }
+      cls_rgw_gc_list_op();
```
Comment (yuvalif) [inline src/rgw/rgw_gc.cc:206]: is this a new function?
where is it defined?

File: src/rgw/rgw_lc.cc [hunk 4]
```diff
@@ -2032,14 +1968,15 @@ int RGWLC::set_bucket_config(RGWBucketInfo& bucket_info,
 
   rgw_bucket& bucket = bucket_info.bucket;
 
-
-  ret = guard_lc_modify(store, bucket, cookie,
+  cls_rgw_lc_set_entry_op();
+  return 0;
+/*  ret = guard_lc_modify(store, bucket, cookie,
```
Comment (yuvalif) [inline src/rgw/rgw_lc.cc:1973]: why is this code commented?

**Review by yuvalif — CHANGES_REQUESTED** [review]
not clear where did all the logic go?
where are the new function defined?

**Comment by dorindabassey** [comment]
Hi Yuval,

I realized what I was supposed to do regarding the PR, so I made updates on
it here https://github.com/ceph/ceph/pull/38124
Thanks.

**Pull request closed without merging**